test/zdtm/static/child_subreaper_and_reparent.c

```diff
@@ -19,132 +19,120 @@ enum {
+}
+
+int subreaper() {
 	int pid, ret, status;
 
-	switch (type) {
-		case MAIN:
-			setsid();
-
-			pid = fork();
-			if (pid < 0) {
-				pr_perror("Failed to fork");
-				exit(1);
-			} else if (pid == 0) {
-				exit(worker(SUBREAPER));
-			}
-
-			/* Wait until ORPHAN is ready to C/R */
-			futex_wait_until(fstate, TEST_CRIU);
-
-			test_daemon();
-			test_waitsig();
-
-			/* Give controll to ORPHAN to check it's parent */
-			futex_set_and_wake(fstate, TEST_CHECK);
-			futex_wait_until(fstate, TEST_EXIT);
-
-			/* Cleanup */
-			while (wait(&status) > 0) {
-				if (!WIFEXITED(status) || WEXITSTATUS(status)) {
-					fail("Wrong exit status: %d", status);
-					return 1;
-				}
-			}
-
-			if (*parent_before_cr != *parent_after_cr)
-				fail("Parent missmatch before %d after %d", *parent_before_cr, *parent_after_cr);
-			else
-				pass();
-			return 0;
-		case SUBREAPER:
-			setsid();
-
-			pid = fork();
-			if (pid < 0) {
-				pr_perror("Failed to fork");
-				return 1;
-			} else if (pid == 0) {
-				exit(worker(HELPER));
-			}
-
-			/* Reap the HELPER */
-			waitpid(pid, &status, 0);
-			if (!WIFEXITED(status) || WEXITSTATUS(status)) {
-				pr_perror("Wrong exit status for helper: %d", status);
-				return 1;
-			}
-
-			ret = prctl(PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0);
-			if (ret) {
-				pr_perror("Can't set child subreaper attribute, err = %d", ret);
-				return 1;
-			}
-
-			/* Give controll to ORPHAN to save it's parent */
-			futex_set_and_wake(fstate, TEST_SAVE);
-			futex_wait_until(fstate, TEST_EXIT);
-			return 0;
-		case HELPER:
-			pid = fork();
-			if (pid < 0) {
-				pr_perror("Failed to fork");
-				return 1;
-			} else if (pid == 0) {
-				exit(worker(ORPHAN));
-			}
-			return 0;
-		case ORPHAN:
-			/*
-			 * Wait until reparented to the pidns init. (By waiting
-			 * for the SUBREAPER to reap our parent.)
-			 */
-			futex_wait_until(fstate, TEST_SAVE);
-
-			*parent_before_cr = getppid();
-
-			/* Return the controll back to MAIN worker to do C/R */
-			futex_set_and_wake(fstate, TEST_CRIU);
-			futex_wait_until(fstate, TEST_CHECK);
-
-			*parent_after_cr = getppid();
-
-			futex_set_and_wake(fstate, TEST_EXIT);
-			return 0;
+	setsid();
+
+	pid = fork();
+	if (pid < 0) {
+		pr_perror("Failed to fork");
+		return 1;
+	} else if (pid == 0) {
+		exit(helper());
 	}
 
+	/* Reap the HELPER */
+	waitpid(pid, &status, 0);
+	if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+		pr_perror("Wrong exit status for helper: %d", status);
+		return 1;
+	}
+
+	ret = prctl(PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0);
+	if (ret) {
+		pr_perror("Can't set child subreaper attribute, err = %d", ret);
+		return 1;
+	}
+
+	/* Give control to ORPHAN to save it's parent */
+	futex_set_and_wake(&sh->fstate, TEST_SAVE);
+	futex_wait_until(&sh->fstate, TEST_EXIT);
 	return 0;
 }
 
-int main(int argc, char **argv)
-{
-	void *ptr;
-
-	BUG_ON(sizeof(*fstate) + sizeof(*parent_before_cr) + sizeof(*parent_after_cr) > 4096);
+int main(int argc, char **argv) {
```

Comment:
We follow linux coding style. Please place beginning brace at the new line for functions.
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces